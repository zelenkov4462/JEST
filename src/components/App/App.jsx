import './App.css';
import {useEffect, useState} from "react";

function App() {
    const [data, setData] = useState(null)
    const [toggle, setToggle] = useState(false)
    const [value, setValue] = useState('')

    const onClick = () => setToggle(prev => !prev)

    useEffect(() => {
        const timer = setTimeout(() => {
            setData({})
        }, 100)
        return () => {
            clearTimeout(timer)
        }
    }, [])
  return (
    <div className="App">
        <div data-testid='value-elem'>{value}</div>
        {toggle && <div data-testid='toggle-elem'>toggle</div>}
        {data && <div style={{color: 'red'}}>data</div>}
      <h1>Hello</h1>
      <button data-testid='toggle-btn' onClick={onClick}>
        click me
      </button>
      <input onChange={(e) => setValue(e.target.value)} type="text"
             placeholder='value...'/>
    </div>
  );
}

export default App;
