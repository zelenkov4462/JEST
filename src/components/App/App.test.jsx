import { render, screen, fireEvent } from "@testing-library/react";

import userEvent from "@testing-library/user-event";

// fireEvent - это объект, предназначенный
// для работы с событиями

import App from "./App";

//Методы -

// find используется для работы с асинхронным кодом

// findBy...- возвращает один элемент, findAll - массив

// getBy..., getAll - обязательно должен найтись элемент
// на странице, иначе выбрасывается ошибка и тест падает

// queryBy..., queryAll - возможность убедиться,
// что элемента нет на страницу

describe("TEST APP", () => {
  test("renders learn react link", () => {
    render(<App />);
    const helloElement = screen.getByText(/hello/i);
    const btn = screen.getByRole("button");
    const input = screen.getByPlaceholderText(/value/i);
    //.toBeInTheDocument - проверка появился наш элемент
    // в дереве DOM или нет
    expect(helloElement).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
    expect(input).toBeInTheDocument();

    // Для отладки теста - .debug()
    // screen.debug()

    expect(input).toMatchSnapshot();
  });

  test("query test", () => {
    render(<App />);
    const elementDOM = screen.queryByText(/hello world/i);
    expect(elementDOM).toBeNull();
  });

  test("find test", async () => {
    render(<App />);
    const elementDOM = await screen.findByText(/data/i);
    expect(elementDOM).toBeInTheDocument();
    expect(elementDOM).toHaveStyle({ color: "red" });
    // screen.debug()
  });

  test("CLICK EVENT", () => {
    render(<App />);
    const btn = screen.getByTestId("toggle-btn");
    expect(screen.queryByTestId("toggle-elem")).toBeNull();

    // fireEvent - смотри выше - нужен для событий,
    // т.к. у btn нет никаких свойств,
    // а мы проверяем на Click

    fireEvent.click(btn);
    //После нажатия на кнопку проверяем,
    // что элемент появился на странице

    expect(screen.queryByTestId("toggle-elem")).toBeInTheDocument();
    fireEvent.click(btn);
    expect(screen.queryByTestId("toggle-elem")).toBeNull();
  });

  test("INPUT TEST", () => {
    render(<App />);
    const input = screen.getByPlaceholderText(/value/i);
    // expect(screen.queryByTestId('value-elem')).toContainHTML('')
    expect(screen.getByTestId("value-elem")).toContainHTML("");

    //fireEvent - это искуственное взаимодействие
    // с интерактивными элементами на странице

    fireEvent.input(input, {
      target: { value: "12345" },
    });

    // expect(screen.queryByTestId("value-elem")).toContainHTML("12345");
    expect(screen.getByTestId("value-elem")).toContainHTML("12345");
  });

  // используем userEvent вместо fireEvent -
  // т.к. полностью расширяет свойства для элементов,
  // с которыми взаимодействует пользователь

  // Отличие - userEvent - не работает с
  // конкретными событиями, а воспроизводит
  // действия пользователя
  // (нажатия кнопок, движения мыши и т.д.)

  test("INPUT TEST with userEvent", () => {
    render(<App />);
    const input = screen.getByPlaceholderText(/value/i);
    expect(screen.getByTestId("value-elem")).toContainHTML("");

    userEvent.type(input, "12345");

    expect(screen.getByTestId("value-elem")).toContainHTML("12345");
  });
});
