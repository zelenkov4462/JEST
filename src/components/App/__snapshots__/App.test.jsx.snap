// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST APP renders learn react link 1`] = `
<input
  placeholder="value..."
  type="text"
/>
`;

exports[`renders learn react link 1`] = `
<input
  placeholder="value..."
  type="text"
/>
`;
